src/openrct2-win/openrct2-win.cpp

```diff
@@ -20,37 +20,86 @@
 #include <iterator>
 #include <openrct2-ui/Ui.h>
 #include <openrct2/core/String.hpp>
+#include <shellapi.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string>
 #include <vector>
 
-static std::vector<std::string> GetCommandLineArgs(int argc, wchar_t** argvW);
+static std::vector<std::string> GetCommandLineArgs(const wchar_t* cmdLine);
+static bool CreateConsole();
 
 /**
- * Windows entry point to OpenRCT2 with a console window using a traditional C main function.
+ * Windows entry point to OpenRCT2 for a GUI application, calling to the traditional C main function.
  */
-int wmain(int argc, wchar_t** argvW, [[maybe_unused]] wchar_t* envp)
+int WINAPI WinMain(
+    [[maybe_unused]] HINSTANCE hInstance, [[maybe_unused]] HINSTANCE hPrevInstance, [[maybe_unused]] LPSTR lpCmdLine,
+    [[maybe_unused]] int nShowCmd)
 {
-    auto argvStrings = GetCommandLineArgs(argc, argvW);
+    auto argvStrings = GetCommandLineArgs(GetCommandLineW());
 
-    SetConsoleCP(CODE_PAGE::CP_UTF8);
-    SetConsoleOutputCP(CODE_PAGE::CP_UTF8);
+    // Handle a special Windows-only commandline argument "--console". If present, attach to the parent process console or
+    // create our own, and point stdin/stdout at that.
+#ifdef DEBUG
+    constexpr bool alwaysCreateConsole = true;
+#else
+    constexpr bool alwaysCreateConsole = false;
+#endif
+
+    auto consoleArg = std::find(argvStrings.begin(), argvStrings.end(), "--console");
+    if (consoleArg != argvStrings.end())
+    {
+        CreateConsole();
+
+        // Remove this argument so the game doesn't throw an unknown argument error.
+        argvStrings.erase(consoleArg);
+    }
+    else if constexpr (alwaysCreateConsole)
+    {
+        CreateConsole();
+    }
 
     std::vector<const char*> argv;
+    argv.reserve(argvStrings.size() + 1);
     std::transform(
         argvStrings.begin(), argvStrings.end(), std::back_inserter(argv), [](const auto& string) { return string.c_str(); });
-    auto exitCode = NormalisedMain(argc, argv.data());
+    argv.push_back(nullptr);
+    auto exitCode = main(static_cast<int>(argvStrings.size()), argv.data());
     return exitCode;
 }
 
-static std::vector<std::string> GetCommandLineArgs(int argc, wchar_t** argvW)
+static std::vector<std::string> GetCommandLineArgs(const wchar_t* cmdLine)
 {
     // Allocate UTF-8 strings
     std::vector<std::string> argv;
-    for (int i = 0; i < argc; i++)
+    int numArgs;
+    LPWSTR* args = CommandLineToArgvW(cmdLine, &numArgs);
+    if (args != nullptr)
     {
-        argv.push_back(String::ToUtf8(argvW[i]));
+        argv.reserve(numArgs);
+        for (int i = 0; i < numArgs; i++)
+        {
+            argv.push_back(String::ToUtf8(args[i]));
+        }
     }
+    LocalFree(args);
     return argv;
 }
+
+static bool CreateConsole()
```

Comment:
Function results a value but is not used. What's the intent?

Reply:
It was used previously to free the console, but that caused an exit crash in `StdInOutConsole` so I removed it - the process is tearing down anyway. I personally don't see much harm in keeping this return value, but I can remove it if needed.

Reply:
Discarding the result of a function feels bad